README.md
# Control Philips Hue Light

# Setup

# Usage
This package enables users to change lights to change their mood. For example, to encourage the user, do the following:
```
from control_huelight_python import control
control.set_encouragement_mood()
```
Availabilities
-[x] control.set_encouragement_mood()
-[x] control.set_fresh_mood()
-[x] control.set_super_performance_mood()

# Color Temperature Lookup Table
To enable the capability to set a typical LED color while changing the temperature, I use the HSV color space and find the range of H value for ordinary colors i.e. red, yellow, green … This also allows us to change the brightness by changing the V value. Depending on which color space an LED manufacture uses, we can convert from the HSV space to that space. For example, Phillips Hue light bulbs use XY color space. To determine the range of HSV values for each color, I used the following scale
![color scale](media/hsv_color_scale.gif)

Changelog.md
# Change Log
All major changes are listed here.

# [v0.1.1] Nov 10th 2021
## Added


## Changes
- [] Reorganize the Hue Light and Controller classes

tests/test_version.py
from control_huelight_python import __version__


def test_version():
    assert __version__ == "0.1.1"


tests/test_setup_devices.py
"""
@Brief: test device setup
"""

from control_huelight_python.devices import HueLight, HueBridgeController
import logging

logging.basicConfig(level=logging.DEBUG)


def test_setup_devices():
    """
    @Brief: test device setup
    """
    huelight1 = HueLight(name="Hue color lamp bedroom 1")
    huelight2 = HueLight(name="Hue color lamp 2")
    huecontroller = HueBridgeController(None, None)

    # Test the device connection
    is_connected = huecontroller.check_light_connection(huelight1)
    assert is_connected is True
    is_connected = huecontroller.check_light_connection(huelight2)
    assert is_connected is True


if __name__ == "__main__":
    test_setup_devices()


tests/test_control_huelight.py
import logging
from control_huelight_python import control
import time


def test_control_huelight():
    # Send encouragement to user
    logging.info("Setting Encourage mood...")
    control.set_encouragement_mood()
    logging.info("Set Encourage mood. Sleep for 5 seconds")
    time.sleep(5)
    logging.info("Setting Speedup mood...")
    control.set_super_performance_mood()
    logging.info("Set Speedup mood. Sleep for 5 seconds")
    time.sleep(5)
    control.set_fresh_mood()
    logging.info("Set fresh mood. Sleep for 5 seconds")
    time.sleep(5)
    logging.info("Completed tests")
    # control.set_modern_mood()
    # logging.info("Set modern mood. Sleep for 5 seconds")
    # time.sleep(5)


if __name__ == "__main__":
    test_control_huelight()


control_huelight_python/control.py
from phue import Bridge
import logging
from control_huelight_python.utils import color_utils

# from color_utils import MOOD_BOOSTING_COLOR_MAP, COLOR_RGB_MAP

logging.basicConfig(level=logging.INFO)


class BedroomHueLight(object):
    # user_name = "IIvz5qbqEzvJt-0pc6Xt4QAWVt5Od0rbsnzrUIYJ"
    light_unique_id = "00:17:88:01:06:c6:ed:f5-0b"
    light_name = "Hue color lamp bedroom 1"
    light_id = 5


def login_hue_bridge(
    bridge_ip="192.168.1.15", user_name="x-kHU4Ng800tNMpNIC51mF3l6V1bL9w4AuNB7-pI"
):
    bridge = Bridge(bridge_ip)
    return bridge


class BedroomHuelightController(object):
    def __init__(self, bedroom_hue_light_name):
        self.bridge = login_hue_bridge()
        # self.bridge.connect()
        self.light_name = bedroom_hue_light_name
        self.light_id = self.get_light_id()

    def get_light_id(self):
        for light in self.bridge.lights:
            if light.name == self.light_name:
                return light.light_id

    def turn_on(self):
        for light in self.bridge.lights:
            if light.name == self.light_name:
                light.on = True

    def turn_off(self):
        for light in self.bridge.lights:
            if light.name == self.light_name:
                light.on = False

    def set_brightness(self, brightness):
        # Transition: 5 seconds
        command = {"transitiontime": 50, "on": True, "bri": brightness}
        self.bridge.set_light(self.light_id, command)

    def set_color(self, xy):
        command = {"transitiontime": 50, "xy": xy}
        self.bridge.set_light(self.light_id, command)


def set_light_for_a_mood(mood_name: str = "encourage"):
    """Change the color of the light to the color of the mood
    Available moods: encourage, force_speedup, calm, fresh
    """
    # Get the color of the mood
    mood_colors = color_utils.get_mood_color(mood_name)
    if isinstance(mood_colors, list):
        mood_color = mood_colors[0]
    else:
        mood_color = mood_colors
    mood_color_rgb = color_utils.COLOR_RGB_MAP[mood_color]
    mood_color_xy = color_utils.rgb_to_xy(mood_color_rgb)

    # Set the color of the light
    light = BedroomHueLight()
    controller = BedroomHuelightController(light.light_name)
    logging.info(
        f"Set the color of the light to {mood_name}. Color: {mood_color} | xy: {mood_color_xy}"
    )
    controller.set_brightness(100)
    controller.set_color(mood_color_xy)


def set_encouragement_mood():
    """Change the color to encourage the user's mood
    # Set the color to the color of the mood
    """
    mood_name = "encourage"
    set_light_for_a_mood(mood_name)


def set_super_performance_mood():
    """Change the color to speed up the working speed
    """
    mood_name = "force_speedup"
    set_light_for_a_mood(mood_name)


def set_modern_mood():
    """Change the color to refresh the user's mood
    """
    mood_name = "modern"
    set_light_for_a_mood(mood_name)


def set_fresh_mood():
    """Change the color to refresh the user's mood
    """
    mood_name = "fresh"
    set_light_for_a_mood(mood_name)


if __name__ == "__main__":
    light = BedroomHueLight()
    controller = BedroomHuelightController(light.light_name)
    # controller.turn_on()
    controller.set_brightness(50)
    controller.set_color([0.1, 0.2])


control_huelight_python/devices/devices.py
from typing import Tuple, List, Union
from phue import Bridge as phue_Bridge
from control_huelight_python.utils import color_utils
import logging


class HueLight(object):
    def __init__(self, name: str = "Hue color lamp bedroom 1", id: int = 5) -> None:
        """ Initialize the HueLight object
        Args:
            name (str, optional): Name of the light that appear on the Hue app. Defaults to "Hue color lamp bedroom 1".
            id (int, optional): ID of the light defined by the Hue bridge. Defaults to 5.
        """
        super().__init__()
        self.name = name
        self.id = id
        self.phue_light = None  # Light() object from phue package


class HueBridgeController(object):
    def __init__(
        self,
        bridge_ip: str = "192.168.1.15",
        user_name="x-kHU4Ng800tNMpNIC51mF3l6V1bL9w4AuNB7-pI",
    ) -> None:
        """ Initialize the a Hue Bridge controller for multiple connected devices
        Args:
            bridge_ip (str, optional): IP address of the Hue bridge. Defaults to "192.168.1.15"
            user_name (str, optional): Your user name to log into the Hue bridge. Defaults to "x-kHU4Ng800tNMpNIC51mF3l6V1bL9w4AuNB7-pI".
        """
        super().__init__()
        self.bridge = phue_Bridge(bridge_ip, user_name)
        try:
            self.bridge.connect()
        except RuntimeError as e:
            print("Cannot connect")
            logging.error(e)
            return None
        # Connected lights that are already verified by the user
        self.verified_light_ids = []

    def check_light_connection(self, light: HueLight) -> bool:
        """ Check if the light is connected to the bridge. It will also fulfill missing information about the light
        Args:
            light (HueLight): Light to check
        Returns:
            bool: True if the light is connected to the bridge, False otherwise
        """
        connected = False
        if light.name is not None:
            for light_obj in self.bridge.lights:
                logging.debug(f"-> available light: {light_obj.name}")
                if light.name == light_obj.name:
                    light.id = light_obj.light_id
                    light.phue_light = light_obj
                    connected = True
                    self.verified_light_ids.append(light.id)
                    logging.debug(f"Light {light.name} is connected")
                    break
        elif light.id is not None:
            for light_obj in self.bridge.lights:
                if light.id == light_obj.light_id:
                    light.name = light_obj.name
                    light.phue_light = light_obj
                    connected = True
                    self.verified_light_ids.append(light.id)
                    logging.debug(f"Light {light.name} is connected")
                    break
        else:
            logging.error("Light must be given with either name or id. None is given!")
        return connected

    def set_light_temperature(
        self,
        light: HueLight,
        temp: int,
        brightness: int = 254,
        transition_time: int = 30,
    ) -> None:
        """ Set the temperature of a light
        Args:
            light (HueLight): Light to set
            temp (int): Color temperature in [154, 500]
            brightness (int, optional): Brightness of the light in [0, 254]. Defaults to 254.
            transition_time (int, optional): Transition time in [0, 65535]. Defaults to 3 seconds
        """
        if not self.check_light_connection(light):
            raise RuntimeError(f"Light {light.name} is not connected")

        logging.debug(f"Current light temperature: {light.phue_light.colortemp}")
        command = {
            "transitiontime": transition_time,
            "on": True,
            "ct": temp,
            "bri": brightness,
        }
        self.bridge.set_light(light.id, command)
        logging.debug(f"Set light temperature: {light.phue_light.colortemp}")

    def turn_light_on(self, light: HueLight) -> None:
        """ Turn on a light
        Args:
            light (HueLight): Light to turn on
        """
        if not self.check_light_connection(light):
            raise RuntimeError(f"Light {light.name} is not connected")
        light.phue_light.on = True

    def turn_light_off(self, light: HueLight) -> None:
        """ Turn off a light
        Args:
            light (HueLight): Light to turn off
        """
        if not self.check_light_connection(light):
            raise RuntimeError(f"Light {light.name} is not connected")
        light.phue_light.on = False

    def set_light_brightness(self, light: HueLight, brightness):
        """
        Set the brightness of a light
        Args:
            light (HueLight): Light to set
            brightness (int): Brightness of the light in [0, 254].
        """
        if not self.check_light_connection(light):
            raise RuntimeError(f"Light {light.name} is not connected")
        if not light.phue_light.on:
            light.phue_light.on = True
        # Transition: 5 seconds
        command = {"transitiontime": 50, "on": True, "bri": brightness}
        self.bridge.set_light(light.id, command)

    def increase_light_brightness(self, light: HueLight, step: int = 12) -> None:
        """
        Increase the brightness of a light by a step
        Args:
            light (HueLight): Light to increase
            step (int, optional): Step to increase the brightness. Defaults to 12.
        """
        if not self.check_light_connection(light):
            raise RuntimeError(f"Light {light.name} is not connected")
        if not light.phue_light.on:
            light.phue_light.on = True

        if light.phue_light.brightness + step >= 254:
            light.phue_light.brightness = 254
        else:
            light.phue_light.brightness += step

    def set_light_xy_color(
        self, light: HueLight, xy: Tuple, transition_time: int = 30
    ) -> None:
        """
        Set the color of a light in xy color space
        Args:
            xy (Tuple): Color in xy color space
        """
        if not self.check_light_connection(light):
            raise RuntimeError(f"Light {light.name} is not connected")
        if not light.phue_light.on:
            light.phue_light.on = True
        command = {"transitiontime": transition_time, "xy": xy}
        self.bridge.set_light(light.id, command)

    def set_light_rgb_color(
        self, light: HueLight, rgb: Tuple, transition_time: int = 30
    ) -> None:
        """
        Set the color of a light in rgb color space
        Args:
            rgb (tuple): Color in rgb color space
        """
        if not self.check_light_connection(light):
            raise RuntimeError(f"Light {light.name} is not connected")
        if not light.phue_light.on:
            light.phue_light.on = True

        if isinstance(rgb, str):
            rgb = color_utils.str_to_rgb(rgb)
        xy = color_utils.rgb_to_xy(rgb)
        command = {"transitiontime": transition_time, "xy": xy}
        self.bridge.set_light(light.id, command)


control_huelight_python/utils/color_utils.py
import logging

logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)


HSV_THRESH_VALUES = {
    "red": {
        "h_low": -15,
        "h_high": 30,
        "s_low": 0.60,
        "s_high": 1.0,
        "cool_to_warm": True,
    },
    "orange": {
        "h_low": 30,
        "h_high": 45,
        "s_low": 0.60,
        "s_high": 1.0,
        "cool_to_warm": True,
    },
    "yellow": {
        "h_low": 45,
        "h_high": 90,
        "s_low": 0.60,
        "s_high": 1.0,
        "cool_to_warm": False,
    },
    "green": {
        "h_low": 90,
        "h_high": 150,
        "s_low": 0.60,
        "s_high": 1.0,
        "cool_to_warm": False,
    },
    "cyan": {
        "h_low": 150,
        "h_high": 225,
        "s_low": 0.60,
        "s_high": 1.0,
        "cool_to_warm": False,
    },
    "blue": {
        "h_low": 225,  # 225. Add 20 to get very blue
        "h_high": 270,  # 270. -10 to get very blue
        "s_low": 0.60,
        "s_high": 1.0,
        "cool_to_warm": True,
    },
    "violet": {
        "h_low": 270,
        "h_high": 285,
        "s_low": 0.60,
        "s_high": 1.0,
        "cool_to_warm": True,
    },
    "magenta": {
        "h_low": 285,
        "h_high": 345,
        "s_low": 0.60,
        "s_high": 1.0,
        "cool_to_warm": True,
    },
}


COLOR_RGB_MAP = {
    "black": (0, 0, 0),
    "white": (255, 255, 255),
    "red": (255, 0, 0),
    "green": (0, 255, 0),
    "blue": (0, 0, 255),
    "yellow": (255, 255, 0),
    "orange": (255, 165, 0),
    "purple": (128, 0, 128),
    "pink": (255, 192, 203),
    "grey": (128, 128, 128),
    "gray": (128, 128, 128),
    "silver": (192, 192, 192),
    "brown": (165, 42, 42),
    "maroon": (128, 0, 0),
    "lime": (0, 255, 0),
    "teal": (0, 128, 128),
    "navy": (0, 0, 128),
    "olive": (128, 128, 0),
    "fuchsia": (255, 0, 255),
    "aqua": (0, 255, 255),
    "violet": (238, 130, 238),
    "indigo": (75, 0, 130),
    "gold": (255, 215, 0),
    "silver": (192, 192, 192),
    "coral": (255, 127, 80),
    "salmon": (250, 128, 114),
    "plum": (221, 160, 221),
    "ivory": (255, 255, 240),
    "honeydew": (240, 255, 240),
    "mint": (245, 255, 250),
    "azure": (240, 255, 255),
    "lavender": (230, 230, 250),
    "beige": (245, 245, 220),
    "mintcream": (245, 255, 250),
    "whitesmoke": (245, 245, 245),
    "ghostwhite": (248, 248, 255),
    "aliceblue": (240, 248, 255),
    "cyan": (0, 255, 255),
    "lightcyan": (224, 255, 255),
    "mintcream": (245, 255, 250),
    "lavenderblush": (255, 240, 245),
}

MOOD_BOOSTING_COLOR_MAP = {
    "force_speedup": "red",
    "force_slowdown": "blue",
    "fresh": "white",
    "modern": "silver",
    "calm": ["blue"],
    "encourage": ["blue"],
}


def str_to_rgb(color):
    """
    Convert string to RGB color.
    """
    if color in COLOR_RGB_MAP:
        return COLOR_RGB_MAP[color]
    else:
        logger.debug(f"List of available colors: \n {COLOR_RGB_MAP.keys()}")
        logging.debug(f"No valid color is chosen. Return 'white' by default.")
        return COLOR_RGB_MAP["white"]


def get_mood_color(mood):
    """
    Get appropriate color for mood based on the moode description
    @Args:
        mood: mood description
    """
    if mood in MOOD_BOOSTING_COLOR_MAP:
        return MOOD_BOOSTING_COLOR_MAP[mood]
    else:
        logging.debug(f"List of available moods: \n {MOOD_BOOSTING_COLOR_MAP.keys()}")
        logging.debug(f"No valid color mood is chosen. Return 'encourage' by default.")
        return MOOD_BOOSTING_COLOR_MAP["encourage"]


def rgb_to_hsv(*rgb):
    """
    Convert RGB to HSV.
    """
    if len(rgb) == 3:
        r, g, b = rgb
    elif len(rgb) == 1:
        r, g, b = rgb[0]
    r, g, b = r / 255.0, g / 255.0, b / 255.0
    mx = max(r, g, b)
    mn = min(r, g, b)
    df = mx - mn
    if mx == mn:
        h = 0
    elif mx == r:
        delta = (g - b) / df
        h = 60 * (delta % 6)
    elif mx == g:
        delta = (b - r) / df
        h = 60 * (delta + 2)
    elif mx == b:
        delta = (r - g) / df
        h = 60 * (delta + 4)
    if mx == 0:
        s = 0
    else:
        s = df / mx
    v = mx
    return {"h": int(h), "s": s, "v": v}


def hsv_to_rgb(*hsv):
    """ Change HSV to RGB
        @Args:
            hsv: hue, saturation, value
    """
    if len(hsv) == 3:
        h, s, v = hsv
    elif len(hsv) == 1:
        h, s, v = hsv[0]
    h = h / 60.0
    i = int(h)
    f = h - i
    p = v * (1 - s)
    q = v * (1 - s * f)
    t = v * (1 - s * (1 - f))
    if i == 0:
        r, g, b = v, t, p
    elif i == 1:
        r, g, b = q, v, p
    elif i == 2:
        r, g, b = p, v, t
    elif i == 3:
        r, g, b = p, q, v
    elif i == 4:
        r, g, b = t, p, v
    elif i == 5:
        r, g, b = v, p, q
    return int(r * 255), int(g * 255), int(b * 255)


def pick_hsv_color(
    color: str, warmness_percentage: float = 1, brightness_percentage: float = 1
):
    """
    Pick a color with warmness level from the HSV color space (H is the hue, S is the saturation, V is the value).
    H in [0, 360], S in [0, 1], V in [0, 1]
    @Args:
        color: color name
        warmness_percentage: percentage of warmness in [0, 1]
        brightness_percentage: percentage of brightness in [0, 1]
    """
    if color not in HSV_THRESH_VALUES:
        raise ValueError(f"No HSV threshold values for color {color}")
    if warmness_percentage < 0 or warmness_percentage > 1:
        raise ValueError(f"Warmness percentage should be in [0, 1]")

    h_low = HSV_THRESH_VALUES[color]["h_low"]
    h_high = HSV_THRESH_VALUES[color]["h_high"]
    s_low = HSV_THRESH_VALUES[color]["s_low"]
    s_high = HSV_THRESH_VALUES[color]["s_high"]
    v_low = 0
    v_high = 1

    # Increase s_low by 0.1 to avoid white color
    s_low += 0.1

    # The warmness increases as the hue increases?
    cool_to_warm = HSV_THRESH_VALUES[color]["cool_to_warm"]
    if not cool_to_warm:
        warmness_percentage = 1 - warmness_percentage

    # Calculate v value
    v = v_low + (v_high - v_low) * brightness_percentage

    # Calculate h value
    h = h_low + (h_high - h_low) * warmness_percentage
    h = int(h + 360) % 360  # Make sure h is in [0, 360]

    # Calculate s value
    s = s_low + (s_high - s_low) * warmness_percentage
    return [h, s, v]


def pick_rgb_color_by_name(
    color: str, warmness_percentage: float = 1, brightness_percentage: float = 1
):
    """
    Pick a color with warmness level by name.
    @Args:
        color: color name
        warmness_percentage: percentage of warmness in [0, 1]
        brightness_percentage: percentage of brightness in [0, 1]
    """
    # Use the HSV color space to pick a color
    hsv = pick_hsv_color(color, warmness_percentage, brightness_percentage)
    # Convert HSV to RGB
    rgb = hsv_to_rgb(*hsv)
    return rgb


def rgb_to_xy(*rgb):
    """ conversion of RGB colors to CIE1931 XY colors
    Formulas implemented from: https://gist.github.com/popcorn245/30afa0f98eea1c2fd34d
    Args:
        rbg (list): a list of 3 numbers representing red, green and blue in the RGB space
    Returns:
        xy (list): x and y
    """
    if len(rgb) == 3:
        r, g, b = rgb
    elif len(rgb) == 1:
        r, g, b = rgb[0]
    r, g, b = r / 255.0, g / 255.0, b / 255.0
    red, green, blue = r, g, b
    # gamma correction
    red = pow((red + 0.055) / (1.0 + 0.055), 2.4) if red > 0.04045 else (red / 12.92)
    green = (
        pow((green + 0.055) / (1.0 + 0.055), 2.4)
        if green > 0.04045
        else (green / 12.92)
    )
    blue = (
        pow((blue + 0.055) / (1.0 + 0.055), 2.4) if blue > 0.04045 else (blue / 12.92)
    )

    # convert rgb to xyz
    x = red * 0.649926 + green * 0.103455 + blue * 0.197109
    y = red * 0.234327 + green * 0.743075 + blue * 0.022598
    z = green * 0.053077 + blue * 1.035763

    # convert xyz to xy
    x = x / (x + y + z)
    y = y / (x + y + z)

    # TODO check color gamut if known
    return [x, y]
